mirgecom/logging_quantities.py

```diff
@@ -0,0 +1,247 @@
+
+        from functools import partial
+
+        if op == "min":
+            self._myop = partial(self.discr.nodal_min, "vol")
+        elif op == "max":
+            self._myop = partial(self.discr.nodal_max, "vol")
+        elif op == "sum":
+            self._myop = partial(self.discr.nodal_sum, "vol")
+        else:
+            raise RuntimeError(f"unknown operation {op}")
+
+    @property
+    def default_aggregator(self):
+        """Rank aggregator to use."""
+        if self.rank_aggr == "norm_2":
+            from pytools import norm_2
+            return norm_2
+        elif self.rank_aggr == "norm_inf":
+            from pytools import norm_inf
+            return norm_inf
+        elif self.rank_aggr == "sum":
+            return sum
+        elif self.rank_aggr == "min":
+            return min
+        elif self.rank_aggr == "max":
+            return max
+        else:
+            return None
+
+    def __call__(self):
+        """Return the requested quantity."""
+        raise NotImplementedError
+
+
+class ConservedQuantity(PhysicalQuantity):
+    """Logging support for conserved quantities (mass, energy, momentum)."""
+
+    def __init__(self, discr: Discretization, quantity: str, op: str,
+                 unit: str = None, dim: int = None, name: str = None,
+                 rank_aggr: str = None):
+        if unit is None:
+            if quantity == "mass":
+                unit = "kg"
+            elif quantity == "energy":
+                unit = "J"
+            elif quantity == "momentum":
+                if dim is None:
+                    raise RuntimeError("Missing 'dim' parameter for dimensional "
+                                       f"ConservedQuantity '{quantity}'.")
+                unit = "kg*m/s"
+            else:
+                unit = ""
+
+        if name is None:
+            name = f"{op}_{quantity}{dim}"
+
+        PhysicalQuantity.__init__(self, discr, quantity, unit, op, name, rank_aggr)
+
+        self.dim = dim
+
+    def __call__(self):
+        """Return the requested conserved quantity."""
+        if self.state is None:
+            return None
+
+        from mirgecom.euler import split_conserved
+
+        cv = split_conserved(self.discr.dim, self.state)
+
+        self.state = None
+
+        cq = getattr(cv, self.quantity)
+
+        if not isinstance(cq, DOFArray):
+            return self._myop(cq[self.dim])
+        else:
+            return self._myop(cq)
+
+
+class DependentQuantity(PhysicalQuantity):
+    """Logging support for dependent quantities (temperature, pressure)."""
+
+    def __init__(self, discr: Discretization, eos: GasEOS,
+                 quantity: str, op: str, unit: str = None, name: str = None,
+                 rank_aggr: str = None):
+        if unit is None:
+            if quantity == "temperature":
+                unit = "K"
+            elif quantity == "pressure":
+                unit = "P"
+            else:
+                unit = ""
```

Comment:
Same as above. This code shouldn't need to know about units. It should ask someone else.

Reply:
I don't disagree, but I don't think the functionality is there yet.